### PR TITLE
Test on fully revealed objects, add "natural" distribution

### PR DESCRIPTION
**Pull request by TheSpaceSheep**

### CODE REVIEW

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 1]
```diff
@@ -147,3 +161,7 @@ def update_curriculum_level(self):
             self.curriculum_level += 1
             print('Curriculum level : ', self.curriculum_level)
 
+
+    def train(self, mode=True):
+        self.testing = not mode
+        return super().train(mode)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: You can access this mode via self.training, no need to code this again

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 2]
```diff
@@ -104,23 +116,25 @@ def forward(self, sender_input, labels, receiver_input=None, aux_input=None):
...
-        ).to(device)
-        mask_idxs_to_reveal = (
-                mask_idxs_to_reveal < n_revealed.view(batch_size, -1)
-        ).long()
-        idxs_to_reveal = idxs_to_reveal * mask_idxs_to_reveal
-        idxs_to_reveal = idxs_to_reveal + idxs_to_reveal[:, 0].view(
-            batch_size, 1
-        ).expand(batch_size, self.n_attributes) * (1 - mask_idxs_to_reveal)
+        # only mask attributes when training
+        if not self.testing:
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: I'm not sure if this should be handled this way: Ideally you'd like to retain the ability to test with masks as well (or test with a different number of masks).

Here's what I'd suggest:
1. Quick & dirty solution: have an optional argument: `mask_during_eval` that you can set to True or False
2. More fleshed out solution (maybe in another PR?): Don't change the behaviour here. Instead, create a dedicated callback that implements:
    - `on_validation_begin()`: change the parameters of the `game` object to what you want to evaluate on (number of masks, distribution, etc...)
    - `on_validation_end()`: reset the game parameters to what is used for training

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 3]
```diff
@@ -87,6 +90,15 @@ def forward(self, sender_input, labels, receiver_input=None, aux_input=None):
             probs[:,  -1] = 1.
         elif self.reveal_distribution == 'uniform':
             probs = torch.ones((batch_size, self.curriculum_level)) / self.curriculum_level
+        elif self.reveal_distribution == 'natural':
+            # this distribution simulates sampling from a dataset containing 
+            # every masked input (up to curriculum_level masks)
+
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py:95]: Maybe worth adding a quick line explanining that there are (n choose k) x N_v^k elements with k masks

**Review by pmichel31415 — CHANGES_REQUESTED** [review]
Looks good, two main comments

- No need to define a `testing` variable, `self.training` is native to nn.Modules and does the exact same thing
- It would be better to keep more options for evaluation (see my comment)

File: egg/zoo/compo_vs_generalization/curriculum_games.py (hunk 2 above)
Comment (TheSpaceSheep) [reply egg/zoo/compo_vs_generalization/curriculum_games.py]: Good point, it would also give the possibility to do several different tests with a different number of masks. I will make another PR addressing this.

**Review by pmichel31415 — APPROVED** [review]
LGTM
